cmd/oceantv/broadcast/youtube.go

```diff
@@ -70,6 +74,99 @@ type IDs struct {
 	BID, SID, CID string
 }
 
+// YTConfig stores the state of oauth flows for a youtube authenticaion pipeline.
+type YTConfig struct {
```

Comment:
Where is token getting refreshed?

Unless, I'm missing something, there needs to be code similar to what exists in gauth.GetProfile, I.e.,

```
	// Issue a new client request to refresh the OAuth token.
	ctx := r.Context()
	src := ua.cfg.TokenSource(ctx, tok)
	newTok, err := src.Token()
	if err != nil {
		return nil, fmt.Errorf("could not get refreshed token: %w", err)
	}
	clt := ua.cfg.Client(ctx, newTok)
	
```

Reply:
> Actually, the other methods are all very similar to the methods in cloud/gauth.
> 
> Could we not reuse/extend the latter?

I was definitely using the userauth as inspiration for this code. Currently the user auth defaults the init to the email, and profile scopes, where this change needs youtube scope.

The calback methods and other handlers will also have slightly different implementations, as well as this implementation requires non exported functions from youtube.go. Other than the actual struct being similar, I'm not sure the business logic side of things are easily shared (unless they get factored out)

Does it make sense to use the same struct but different methods on it?

Reply:
> Where is token getting refreshed?
Yes you are correct, good pick up. We will need to check if the current token exists, and if it does, we should refresh, otherwise we should use the code to exchange?

However, I'm not sure how we would determine which token object we need to find, before getting a new token?

```golang
code := r.FormValue("code")
	tok, err := cfg.gConfig.Exchange(context.Background(), code)
	if err != nil {
		log.Printf("could not exchange token: %v", err)
	}
```

Here we are getting the new token, which would allow us to find the user email, to check if the token exists, but at this point we have already got the token. Should this token be used to find the old token, then discarded if it exists, and the old token refreshed? Or is there a more semantic way of handling this?

Reply:
> > Actually, the other methods are all very similar to the methods in cloud/gauth.
> > Could we not reuse/extend the latter?
> 
> I was definitely using the userauth as inspiration for this code. Currently the user auth defaults the init to the email, and profile scopes, where this change needs youtube scope.
> 
> The calback methods and other handlers will also have slightly different implementations, as well as this implementation requires non exported functions from youtube.go. Other than the actual struct being similar, I'm not sure the business logic side of things are easily shared (unless they get factored out)
> 
> Does it make sense to use the same struct but different methods on it?

I think it may be possible to use the same structure, but with optional properties. We could use functional options as we do for the notify package. For example, we `WithGoogleUser `would perform authentication for a Google User, whereas `WithYouTubeAccount` would work on a YouTube account. The OAuth2 flow is almost identical in both cases.

For exampe

Reply:
I believe the pattern is to get the TokenSource from the existing token, then use that to instantiate a new token.